### DISCUSSION TITLE
Printing text on Raspberry Pi - imagemagick error

### DISCUSSION THREAD

**Discussion by johnsreid** [NaitLee/Cat-Printer#29] [Q&A] [answered]
Hey. Amazing library and I am looking forward to hooking the printer up to my Twitch streams for the viewers.

I have everything working via the web UI but can't print text from the command line on my Raspberry Pi (RaspberryOS).

I'm getting the following:
```
Cat Printer
Connecting
Text Printing Mode
Traceback (most recent call last):
  File "/home/raspberry/Downloads/cat-printer/printer.py", line 715, in <module>
    main()
  File "/home/raspberry/Downloads/cat-printer/printer.py", line 695, in main
    _main()
  File "/home/raspberry/Downloads/cat-printer/printer.py", line 668, in _main
    file = magick_text(file, 384,
  File "/home/raspberry/Downloads/cat-printer/printer.py", line 536, in magick_text
    subprocess.Popen(['magick', '-background', 'white', '-fill', 'black',
  File "/usr/lib/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.9/subprocess.py", line 1823, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'magick'
```
I've installed imagemagick via apt so I'm at a loss of where to go next. Is there something obvious that I'm missing?

**Comment by NaitLee** [comment, marked as answer]
Try if any "legacy" IM command would work, e.g. `convert`. Just try it in terminal.  
If that works, you can ~~do `alias magick=convert` in the shell (put to `~/.bashrc` to persist),~~  
or symlink `magick` to `convert`:  
`sudo ln -s /usr/bin/convert /usr/bin/magick`

I will consider circumstance it in the code in the future.

By the way, if what you mean is to print chat content, you may consider using PF2 font technique, as it could stream text buffer *by nature*.
And with it, it can be as simple as piping two commands.
By contrast, IM prints all text in-one-shot, and could be resource heavy, especially in this case (that should be invoked many times)
You can know more from cmdline help or #10.

**Reply by johnsreid** [reply]
Thanks for this. Some really helpful pointers which I hugely appreciate.

The alias method didn't work at all so I tried creating the symlink. The script picks up the symlinked command but it causes additional errors:

```
Cat Printer
Connecting
convert-im6.q16: no decode delegate for this image format `' @ error/constitute.c/ReadImage/575.
convert-im6.q16: no images defined `pbm:-' @ error/convert.c/ConvertImageCommand/3229.
Finished
Disconnecting from printer
```

Then pushes out some margin feed.

I'm definitely going to give the PF2 font technique a go and see how I get on with that in the meantime. Many thanks for the support!

**Reply by NaitLee** [reply]
I don't think you're using Text Printing mode. Please use `-t` parameter.

Some examples:

```bash
# scan for 2 seconds, convert text content with IM, print a text file
python3 printer.py  -s 2  -t 16,DejaVu-Sans mytextfile.txt
# print output of other program with PF2 font, to a GB03 printer
echo "Hello there!" | python3 printer.py  -s 2,GB03  -t 1,unifont,pf2  -
```

Well, maybe I should clarify that `-c text` is for "text-alike images", not text content.
You can also seek for more with cmdline help and #10.

Something sad about Debian (base of RasPiOS) is that its software package is way too old, even with newest 11.*
An interesting fact is almost everyone come to here are using Arch linux or its derivatives.

Oh right, alias won't work because this program doesn't use shell to invoke programs anymore.

**Reply by johnsreid** [reply]
Lovely. I've got a system working now using the PF2 method and it works perfectly.

I now have it connected to my Twitch channel so people can send messages to it by redeeming channel points - I'll be able to try it out in a couple of days!

I'll keep an eye out for font scaling as the text is a little small - though hardly a deal-breaker.

Thanks for your pointers and for the project here!

**Reply by NaitLee** [reply]
Okay. So it means that I should do font-zoom now, for many people want, and maybe Web (browser) PF2 generator.
If you have Linux with command `grub-mkfont` there, you can generate a PF2 font easily, with bigger size.
You could try if it's on RasPiOS. If not there are many distros around, simply pick one.

**Reply by NaitLee** [reply]
A little note: (PF2) with additional parameter `-p flip` you can print text lines in reversed (chronological) order, and for it is flipping direction you could face the printer toward & showcase it.
If language is hard to understand, simply try it and see printer output...

**Reply by johnsreid** [reply]
I see it! I had a little play with it and it could be fun to use for interactive uses.

Thanks for all of this. 👍🏻

**Reply by NaitLee** [reply]
The font zooming (glyph scaling) is in newest 0.5.2. Use something like `-t 2,unifont,pf2`.

I personally feel 2x Unifont is somewhat big... But it couldn't be in the middle.

You could search if `grub` or `grub-mkfont` is there in RasPiOS apt repo.  
`apt search grub`

If that's the case you can install it & make a satisfying font by yourself, with any ttf/otf.
